### DISCUSSION TITLE
Comparison to sfc64

### DISCUSSION THREAD

**Discussion by flyingmutant** [tkaitchuck/Mwc256XXA64#3] [General]
Hey! I've enjoyed your article about makings of the MWC256XXA64 a lot -- it is really great. However, when comparing this generator to sfc64, would it be true to say that we gain bigger period, but lose some speed (due to multiplication) and maybe "streams" feature (which I consider sfc64 to have, but am not too sure)?

**Comment by rkern** [comment]
I personally don't consider `sfc64` to have a "streams" feature. The state space of `sfc64` is, for sure, divided up into a number of disjoint cycles, while MWC256XXA64 only has one, but there is no way to _select_ a cycle or _ensure_ that two streams are in different cycles in the way that `PCG64` can (if you decide to expose that feature to the user).

One other feature that a MWC PRNG has but `sfc64` doesn't is jumpahead, which some people prefer to use to create parallel streams.

**Reply by tylov** [reply]
Please have a look at my **[stc64 PRNG](https://github.com/tylov/STC/blob/master/docs/crandom_api.md)**. It has a the same base structure as `sfc64`, but has 64 bits extra state. The extra bits are used to enable "streams". Simply adding 2 (to the Weyl-sequence increment) represents a "jump" to a new unique period -- guaranteed minimum 2^64 in length (i.e. there are 2^63 number of streams).

A great property of `sfc64` (and `stc64` ) is that it does not use multiplication or rely on 128 bit integers -- in addition to its amazing speed, simple/small implementation and high quality output.

The only code difference between `sfc64` and `stc64` is the output function, which mixes in the extra state and uses an XOR instead of addition. There are no performance penalty (on some platforms it is faster than `sf64`, surprisingly).

`stc64` can be seen as a more general `sfc64`, and therefore the quality should be equal or better. It can be downscaled to 32 and 16 bits (just like `sfc64`), and all variants have been thoroughly tested with PractRand in particular (naturally as `sfc64` is created by its author).

Speed-wise it is almost twice as fast as `pcg64`. Look at the speed-test in this [fork of Lemire's testingRNG](https://github.com/tylov-fork/testingRNG).  I have added `sfc64`, `stc64` and `romu_trio`, which all are among the fastest of the prngs in the test.

Note:  Lemire's original speed-test is actually quite inaccurate, as he uses function pointers to the prngs, and a memory clobbing which does not appear to be working as intended. My fork has fixed this, and it gives a few quite different results.
